### PR TITLE
Bug/2948 wrong max one service path response

### LINKED ISSUE #2948: Wrong "max one service-path allowed for subscriptions" in NGSIv2 subscription operation
Attemping to subscribe using `POST /v2/subscriptions` with multiple service paths, the following error message is returned:

```
< HTTP/1.1 200 OK
< Connection: Keep-Alive
< Content-Length: 184
< Content-Type: application/json
< Fiware-Correlator: 3fab418c-5a4f-11e7-9af0-000c29173617
< Date: Mon, 26 Jun 2017 09:10:07 GMT
< 
{
  "subscribeError" : {
    "errorCode" : {
      "code" : "400",
      "reasonPhrase" : "Bad Request",
      "details" : "max one service-path allowed for subscriptions"
    }
  }
}
```

The semantics of the error are ok. However, the syntax is not correct for a NGSIv2 message:

1. Payload should be based on `error` and `description`.
2. Response code should be 400 and not 200.

### PR DESCRIPTION
**Pull request by arigliano**
This PR fixes #2948 Issue

- It was implemented a new **toJson()** method in SubscriberError, in order to correctly print out the error response, depending on which serviceRoutine version calls either render() or toJson() method.
- Fixed 400 httpStatusCode in both serviceRoutines

### CODE REVIEW

File: src/lib/ngsi/SubscriptionId.cpp [hunk 1]
```diff
@@ -141,6 +141,40 @@ void SubscriptionId::present(const std::string& indent)
 
 /* ****************************************************************************
 *
+* SubscriptionId::toJson -
+*/
+std::string SubscriptionId::toJson(RequestType container, bool comma)
+{
+  std::string xString = string;
```
Comment (fgalan) [inline src/lib/ngsi/SubscriptionId.cpp]: We really choose a really bad name to store id string in Subscriptions.h... :)

NTC (just a bit surprised of my/our self :)

File: src/lib/ngsi/SubscriptionId.h [hunk 2]
```diff
@@ -47,6 +47,7 @@ typedef struct SubscriptionId
   const char*   c_str(void) const;
   bool          isEmpty(void);
   std::string   render(RequestType container,bool comma);
+  std::string   toJson(RequestType container,bool comma);
```
Comment (fgalan) [inline src/lib/ngsi/SubscriptionId.h]: Style: whitespace after ,

(I understand you take render() as reference, but it is also wrongly styled... could you please fix that line also, pls?)

File: src/lib/ngsi10/SubscribeContextResponse.cpp [hunk 3]
```diff
@@ -60,6 +60,30 @@ SubscribeContextResponse::SubscribeContextResponse(StatusCode& errorCode)
 
 /* ****************************************************************************
 *
+* SubscribeContextResponse::toJson -
+*/
+std::string SubscribeContextResponse::toJson(void)
+{
+  std::string out     = "";
```
Comment (fgalan) [inline src/lib/ngsi10/SubscribeContextResponse.cpp]: Style: excesive whitepace here?

**Comment by fgalan** [comment]
Good work! Only minor comments.

Btw, do the branch in this PR pass `make test` and `make style_check`?

File: src/lib/ngsi10/SubscribeContextResponse.cpp (hunk 3 above)
Comment (kzangeli) [reply src/lib/ngsi10/SubscribeContextResponse.cpp]: [ Depends on what comes after. The '=' may be aligned with the '=' on the following line. If not, then yes ]

File: src/lib/ngsi/SubscribeError.cpp [hunk 4]
```diff
@@ -44,6 +44,45 @@ SubscribeError::SubscribeError()
...
+    // NOTE: the subscriptionId must have come from the request.
+    //       If the field is empty, we are in unit tests and I here set it to all zeroes
+    //
+    if (subscriptionId.get() == "")
+    {
+      subscriptionId.set("000000000000000000000000");
+    }
+    out += ",";
+    out += JSON_PROP("affectedItems") + "[" + JSON_STR(subscriptionId.toJson(requestType, true)) + "]";
+ }
```
Comment (kzangeli) [inline src/lib/ngsi/SubscribeError.cpp]: Indentation (one space seems to be missing)

File: src/lib/ngsi10/SubscribeContextResponse.cpp (hunk 3 above)
Comment (fgalan) [reply src/lib/ngsi10/SubscribeContextResponse.cpp]: Correct. In this case neither L66 (above) or L68 (below) has assignation stamente, thus my comment about excesive whitepace.

File: test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. Create subscription for E1 with multiple service paths header
+
```
Comment (kzangeli) [inline test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test]: This empty line to be removed

File: test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+#
+
+echo "01. Create subscription for E1 with multiple service paths header"
+echo "=============================================================="
```
Comment (kzangeli) [inline test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test]: A few more '='s, to be as long as the preceding line

Comment (kzangeli) [reply test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test]: Also, the description is not very correct.
Suggestion:
```
Attempt to create subscription for E1 with multiple service paths header, and see it fail
```

**Comment by kzangeli** [comment]
What Fermin said, good work!

File: src/lib/ngsi/SubscriptionId.cpp (hunk 1 above)
Comment (kzangeli) [reply src/lib/ngsi/SubscriptionId.cpp]: This was probably me, smells KZ ... :-)
I don't find it that bad though :-D

NTC

Comment (fgalan) [reply src/lib/ngsi/SubscriptionId.cpp]: Just for fun... blame shows this:

imagen

Very ancient code, probably from first commit in public repo... The origin of this variable is lost in History :D

Comment (kzangeli) [reply src/lib/ngsi/SubscriptionId.cpp]: 99% sure it's mine :-)

**Comment by fgalan** [comment]
@arigliano no news since a week ago... Are the feedback comments clear? Do you need any extra clarification? Don't hesitate to ask if you need, please.

File: src/lib/ngsi/SubscriptionId.h (hunk 2 above)
Comment (arigliano) [reply src/lib/ngsi/SubscriptionId.h]: Fixed in 7480bca

File: src/lib/ngsi10/SubscribeContextResponse.cpp (hunk 3 above)
Comment (arigliano) [reply src/lib/ngsi10/SubscribeContextResponse.cpp]: Fixed in 7480bca

File: src/lib/ngsi/SubscribeError.cpp (hunk 4 above)
Comment (arigliano) [reply src/lib/ngsi/SubscribeError.cpp]: Fixed in 7480bca

File: src/lib/ngsi/SubscribeError.cpp [hunk 7]
```diff
@@ -44,6 +44,45 @@ SubscribeError::SubscribeError()
...
+  else if ((requestType          == SubscribeContext)           &&
+           (subscriptionId.get() != "000000000000000000000000") &&
+           (subscriptionId.get() != ""))
+  {
+	out += ",";
```
[the remark this answers is not in the document]
Comment (arigliano) [reply src/lib/ngsi/SubscribeError.cpp]: Fixed in 7480bca

File: test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test (hunk 5 above)
Comment (arigliano) [reply test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test]: Fixed in 7480bca

File: test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test (hunk 6 above)
Comment (arigliano) [reply test/functionalTest/cases/2948_wrong_max_one_service_path_response/wrong_max_one_service_path_response.test]: Fixed in 7480bca

**Comment by arigliano** [comment]
Sorry for the late. With 7480bca commit, i've fixed the style issues.
 - ```make test``` passed

 - ```make style_check``` gives me a lot of errors regarding code parts that I did not touch.
Starting in verbose mode,```make style_check_v``` gives me the following:


 ```
style_check.sh: 111032 lines in 682 files in dir

0001 errors of category 'Do not leave a blank line after "public:"'
0001 errors of category 'Found C++ system header after other header'
0002 errors of category 'Extra space after ( in function call'
0003 errors of category 'Found C system header after other header'
0003 errors of category 'Missing space before ('
0003 errors of category 'Streams are highly discouraged'
0004 errors of category 'Single-argument constructors should be marked explicit'
0005 errors of category 'Line contains only semicolon'
0005 errors of category 'Missing spaces around ='
0006 errors of category 'Closing ) should be moved to the previous line'
0006 errors of category 'Include the directory when naming'
0008 errors of category 'Should have a space between // and comment'
0009 errors of category 'already included at'
0010 errors of category 'More than one command on the same line'
0011 errors of category 'Found C system header after C++ system header'
0013 errors of category '#endif line should be'
0013 errors of category '#ifndef header guard has wrong style'
0015 errors of category 'Do not use namespace using-directives'
0019 errors of category 'Blank line at the start of a code block'
0027 errors of category 'Missing space after ,'
0033 errors of category 'Labels should always be indented at least one space'
0038 errors of category 'Blank line at the end of a code block'
0044 errors of category 'Is this a non-const reference'
0045 errors of category 'Add #include'
0047 errors of category 'At least two spaces is best between code and comments'
0108 errors of category 'Tab found; better to use spaces'
0120 errors of category 'Extra space before ( in function call'
0133 errors of category 'Use int16/int64/etc'
0180 errors of category 'Line ends in whitespace'
0213 errors of category 'Weird number of spaces at line-start'
0249 errors of category 'Lines should be <= 120 characters long'

New style-guide errors:
-----------------------------------------------------
src/lib/common/JsonHelper.cpp:77: Extra space before last semicolon. If this should be an empty statement, use { } instead. [whitespace/semicolon] [5] src/lib/common/string.cpp:433: Complex multi-line /*...*/-style comment found. Lint may give bogus warnings. Consider replacing these with //-style comments, with #if 0...#endif, or with more clearly structured multi-line comments. [readability/multiline_comment] [5] src/lib/common/string.cpp:1055: Consider using gmtime_r(...) instead of gmtime(...) for improved thread safety. [runtime/threadsafe_fn] [2] src/lib/common/tag.cpp:191: Extra space before last semicolon. If this should be an empty statement, use { } instead. [whitespace/semicolon] [5] src/lib/ngsiNotify/Notifier.cpp:1: Complex multi-line /*...*/-style comment found. Lint may give bogus warnings. Consider replacing these with //-style comments, with #if 0...#endif, or with more clearly structured multi-line comments. [readability/multiline_comment] [5] src/lib/ngsiNotify/Notifier.cpp:0: One or more unexpected \r (^M) found;better to use only a \n [whitespace/newline] [1] src/lib/rest/httpRequestSend.cpp:0: One or more unexpected \r (^M) found;better to use only a \n [whitespace/newline] [1]
-----------------------------------------------------
style_check.sh: found 1381 errors (982 mayor, 392 minor errors) in 111032 lines of source code in 682 files (1.24% style-guide-incompatibilities)
```

I don't know if it is expected, but the errors in the part of code that I have modified are no longer there

**Comment by kzangeli** [comment]
Try 'make style'.
Speaking from memory: I think you are executing style check on the entire code, all dirs.
Not all dirs are prepared still. Another make target, called 'style' (I think) uses a special script called 'style check from makefile'. Well, something like that :)

**Comment by arigliano** [comment]
Ok thanks. I've performed '''make style''' and everything passed with no mayor error.

**Comment by fgalan** [comment]
LGTM. Thx!

Passing the ball to @kzangeli for final LGTM

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**